lib/pengine/tags.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <glib.h>
+#include <stdbool.h>
+
+#include <crm/common/util.h>
+#include <crm/pengine/internal.h>
+#include <crm/pengine/pe_types.h>
+
+gchar **
+pe__nodes_with_tag(pe_working_set_t *data_set, const char *tag_name) {
+    GHashTableIter iter;
+    gpointer key, value;
+    GPtrArray *arr = NULL;
+    gchar **retval = NULL;
+
+    g_hash_table_iter_init(&iter, data_set->tags);
+
+    while (g_hash_table_iter_next(&iter, &key, &value)) {
+        if (!crm_str_eq((char *) key, tag_name, TRUE)) {
+            continue;
+        }
+
+        /* If we get here, we've found the entry in the tags table for the
+         * requested tag.
+         */
+        if (arr == NULL) {
+            arr = g_ptr_array_new();
+        }
+
+        /* Iterate over the list of node IDs. */
+        for (GListPtr refs = ((pe_tag_t *) value)->refs; refs; refs = refs->next) {
+            /* Find the node that has this ID. */
+            const char *id = (const char *) refs->data;
+            pe_node_t *node = pe_find_node_id(data_set->nodes, id);
+
+            if (!node) {
+                continue;
+            }
+
+            /* Get the uname for the node and add it to the return list. */
+            g_ptr_array_add(arr, strdup((gchar *) node->details->uname));
```

Comment:
i think you meant g_strdup() -- otherwise the cast is unnecessary and the array is char**

Reply:
Done.